### PR TITLE
[DRAFT] Initial support for Version entities in Entity Hub

### PR DESCRIPTION
**Pull request by fabiaserra**
Start the work required to support `Version` entities on the Entity Hub so we will be able to update its status or other attributes.

### CODE REVIEW

**Comment by BigRoy** [comment]
Note that you can mark the PR itself as Draft instead of having to put it in the PR title.
Also, this PR seems to include https://github.com/ynput/ayon-python-api/pull/157 - is that intentional?

**Comment by fabiaserra** [comment]
I will let @iLLiCiTiT comment on how we should proceed with this

**Comment by iLLiCiTiT** [comment]
Version entity lives under product entity. It also does not make much sense to add version without representation entity.

Adding any of that to entity hub also changes how hub operates. As far as I know we don't have any huge need for this feature. To be honest I don't have spare time for that now.

**Comment by fabiaserra** [comment]
> Version entity lives under product entity. It also does not make much sense to add version without representation entity.
> 
> Adding any of that to entity hub also changes how hub operates. As far as I know we don't have any huge need for this feature, and too be honest I don't have spare time for that now. You can try to add it on your own.

How are you planning on adding support to sync statuses of Versions across any of the project management tools? Without that version statuses in AYON are useless because people will be expecting to be able to check both databases for ground truth of status, if the one in AYON is not on sync we won't be using it at all and all the places where they see `Status` in AYON on a version will be distracting and error prone if anyone tries to change it

**Comment by fabiaserra** [comment]
The changes done here https://github.com/ynput/ayon-python-api/pull/158/commits/7e7ae670aae5c5434879725c0d210d1693eed3cc now allow me to sync statuses on the versions in the ayon-shotgrid plugin

File: ayon_api/entity_hub.py [hunk 1]
```diff
@@ -2474,6 +2505,7 @@ def from_entity_data(cls, project, entity_hub):
             folder_types=project["folderTypes"],
             task_types=project["taskTypes"],
             name=project["name"],
+            statuses=project["statuses"],
```
Comment (fabiaserra) [inline ayon_api/entity_hub.py:2508]: I think this was a bug on the existing code missing `statuses` for the `ProjectEntity`

File: ayon_api/entity_hub.py [hunk 2]
```diff
@@ -2474,6 +2505,7 @@ def from_entity_data(cls, project, entity_hub):
...
             attribs=project["ownAttrib"],
```
Comment (fabiaserra) [inline ayon_api/entity_hub.py:2509]: @iLLiCiTiT and is this correct? wouldn't it be `project["attrib"]` instead? and why does `ownAttrib` not work for the version entity?

**Comment by iLLiCiTiT** [comment]
> How are you planning on adding support to sync statuses of Versions across any of the project management tools? 

Well, I guess we never though about syncing versions from other project management. Versions in AYON need representations, representations need specific data, how you'll get representation from project management.

File: ayon_api/entity_hub.py (hunk 2 above)
Comment (iLLiCiTiT) [reply ayon_api/entity_hub.py:2509]: Because you set `own_attributes=False`.

File: ayon_api/entity_hub.py (hunk 1 above)
Comment (iLLiCiTiT) [reply ayon_api/entity_hub.py:2508]: Yes, looks like the method is not used.

File: ayon_api/entity_hub.py (hunk 2 above)
Comment (fabiaserra) [reply ayon_api/entity_hub.py:2509]: Well like I say on the code, if I set it to True it errors out...

**Comment by iLLiCiTiT** [comment]
> To your question on how do you get representantion from project management? In SG there's the published files linked to a version that correspond to the representations so it would be easy to map that. In ftrack it's the components of the version... it would definitely be possible

I'm not talking about files, I'm talking about the rest of the representation entity data.

> This is not about creating the versions that are misssing in either, but syncing the data of already created versions on both systems. When a version gets created in AYON a version gets created in SG with a link to it. If I then update the status of that version in SG we need to update that in AYON as well. This PR allows me to do exactly that.

That still requires `ProductEntity`.

**Comment by fabiaserra** [comment]
> I'm not talking about files, I'm talking about the rest of the representation entity data.
 
Which data? A lot of the data fields are already being set in SG versions, the data that's missing would just be left as default or unset 
 
> That still requires `ProductEntity`.

For what reason? The SG <-> AYON linkage is happening at the version level, I'm updating the version entity directly, what do I need to update on the product?

File: ayon_api/entity_hub.py (hunk 2 above)
Comment (iLLiCiTiT) [reply ayon_api/entity_hub.py:2509]: Aaah, I found out that only projects, folders and tasks have `ownAttrib` will make PR marking the argument deprecated in other methods.

**Comment by iLLiCiTiT** [comment]
> Which data? A lot of the data fields are already being set in SG versions, the data that's missing would just be left as default or unset

Representation entity is not just files, random data would make it unusable, but this PR is about versions, ok.

> For what reason? The SG <-> AYON linkage is happening at the version level, I'm updating the version entity directly, what do I need to update on the product?

I do understand but the entity hub is now missing an entity inbetween folder and version.

File: ayon_api/entity_hub.py (hunk 2 above)
Comment (fabiaserra) [reply ayon_api/entity_hub.py:2509]: Haha and does that make sense to pass it on as the "attribs" argument on the creator? What would be the equivalent for that for the other entities?

File: ayon_api/entity_hub.py [hunk 3]
```diff
@@ -2971,3 +3003,156 @@ def _get_label_value(self):
...
+            value is defined based on value of 'entity_id'.
+    """
+
+    entity_type = "version"
+    parent_entity_types = ["product"]
+
+    def __init__(
+        self,
+        *args,
+        tags=None,
```
Comment (iLLiCiTiT) [inline ayon_api/entity_hub.py:3032]: We should add `task_id`.

**Comment by fabiaserra** [comment]
> Representation entity is not just files, random data would make it unusable, but this PR is about versions, ok.

As long as the id gets created and the paths filled up I don't see why it would be completely unusable (the fact that you say it is it worries me the system is more complex than it needs to be) but anyway, it's not the point of this PR nor my intention yet, this is only to allow syncing ALREADY created versions in both systems.

> I do understand but the entity hub is now missing an entity inbetween folder and version.

It's missing quite a lot more than just that haha we can add the missing things as soon as they are needed. The product entity is not needed at all for the SG integration

**Comment by iLLiCiTiT** [comment]
> It's missing quite a lot more than just that haha we can add the missing things as soon as they are needed. The product entity is not needed at all for the SG integration

I do understand, but that doesn't mean it should not be added.

**Comment by fabiaserra** [comment]
> I do understand, but that doesn't mean it should not be added.

Yeah I agree it should be added, but not necessarily part of this PR

File: ayon_api/entity_hub.py (hunk 3 above)
Comment (fabiaserra) [reply ayon_api/entity_hub.py]: done at https://github.com/ynput/ayon-python-api/pull/158/commits/50e75cdfe634ab773ecf31cd44b9986a0a8b18c0

is `comment` something passed as an `attrib`?

**Comment by iLLiCiTiT** [comment]
> Yeah I agree it should be added, but not necessarily part of this PR

It should be added before `VersionEntity`.

**Comment by fabiaserra** [comment]
> > Yeah I agree it should be added, but not necessarily part of this PR
> 
> It should be added before `VersionEntity`.

Agree to disagree :) it's working fine for me and I'm not missing the product entity at all so I'm not going to work on that to make you guys approve this PR sorry

**Comment by mkolar** [comment]
@iLLiCiTiT we'll soon be starting the work on 2 way full sync for ftrack. Is it worth revisiting this?

**Comment by iLLiCiTiT** [comment]
Closing in favor of https://github.com/ynput/ayon-python-api/pull/197

**Pull request closed without merging**